packages/compiler/core/program.ts

```diff
@@ -56,6 +56,12 @@ import {
   mutate,
 } from "./util.js";
 
+export function BackCompatSupportForCadlMain(packageJson: any) {
+  if (packageJson?.tspMain === undefined && packageJson?.cadlMain !== undefined) {
+    packageJson.tspMain = packageJson.cadlMain;
+  }
+}
```

Comment:
I find it surprising that this mutates. 

I would prefer something like `export function getTspMain(packageJson: any) { return packageJson?.tspMain ?? packageJson?.cadlMain }`

Then replace `pkg?.tspMain` with getTspMain(pkg) throughout.

Also, I would move this to a util file where we don't export it publicly to everyone.